geopandas/tools/sjoin.py

```diff
@@ -329,3 +333,161 @@ def _frame_join(indices, left_df, right_df, how, lsuffix, rsuffix):
             joined.index.name = right_index_name
 
     return joined
+
+
+def _nearest_query(
+    left_df: GeoDataFrame,
+    right_df: GeoDataFrame,
+    max_distance: float,
+    how: str,
+    return_distance: bool,
+):
+    if how == "inner":
+        # use whichever input is largest for the index, and the smallest for querying
+        use_left_as_sindex = len(left_df) > len(right_df)
+    else:
+        # otherwise, use the opposite of the join direction for the index
+        use_left_as_sindex = how == "right"
+    if use_left_as_sindex:
+        sindex = left_df.sindex
+        query = right_df.geometry
+    else:
+        sindex = right_df.sindex
+        query = left_df.geometry
+    if not hasattr(sindex, "nearest_all"):
+        raise AttributeError(
+            f"Expected the spatial index to implement `nearest_all`,"
+            f" but {type(sindex)} does not implement this method."
+            " Currently, only PyGEOS supports `nearest_all`."
+        )
+    if sindex:
+        res = sindex.nearest_all(
+            query, max_distance=max_distance, return_distance=return_distance
+        )
+        if return_distance:
+            (input_idx, tree_idx), distances = res
+        else:
+            (input_idx, tree_idx), distances = res, None
+        if use_left_as_sindex:
+            l_idx, r_idx = tree_idx, input_idx
+            sort_order = np.argsort(l_idx, kind="stable")
+            l_idx, r_idx = l_idx[sort_order], r_idx[sort_order]
+            if distances is not None:
+                distances = distances[sort_order]
+        else:
+            l_idx, r_idx = input_idx, tree_idx
+        join_df = pd.DataFrame(
+            {"_key_left": l_idx, "_key_right": r_idx, "distances": distances}
+        )
+    else:
+        # when sindex is empty / has no valid geometries
+        join_df = pd.DataFrame(
+            columns=["_key_left", "_key_right", "distances"], dtype=float
+        )
+    return join_df
+
+
+def sjoin_nearest(
+    left_df,
+    right_df,
+    how="left",
+    max_distance=None,
+    lsuffix="left",
+    rsuffix="right",
+    distance_col=None,
+) -> GeoDataFrame:
+    """Spatial join of two GeoDataFrames based on the distance between their geometries.
+
+    See the User Guide page :doc:`../../user_guide/mergingdata` for details.
+
```

Comment:
It would be good to add a note that the results will include multiple output records for a single input record where there are multiple equidistant nearest or intersected neighbors.

Reply:
done, thank you for the suggestion